frontend/ui/widget/filechooser.lua

```diff
@@ -159,74 +158,27 @@ function FileChooser.getListItem(f, filename, attributes, collate)
         fullpath = filename,
         attr = attributes,
     }
-    if collate then -- file
+    if collate.can_collate_mixed and attributes.mode == "directory" then
+        if collate.item_func ~= nil then
+            item = collate.item_func(item)
+        end
+    elseif attributes.mode == "file" then
         if G_reader_settings:readSetting("show_file_in_bold") ~= false then
             item.opened = DocSettings:hasSidecarFile(filename)
         end
-        if collate == "type" then
-            item.suffix = util.getFileNameSuffix(f)
-        elseif collate == "percent_unopened_first" or collate == "percent_unopened_last" then
-            local percent_finished
-            item.opened = DocSettings:hasSidecarFile(filename)
-            if item.opened then
-                local doc_settings = DocSettings:open(filename)
-                percent_finished = doc_settings:readSetting("percent_finished")
-            end
-            item.percent_finished = percent_finished or 0
+        if collate.item_func ~= nil then
+            item = collate.item_func(item)
         end
     end
     return item
 end
 
 function FileChooser:getSortingFunction(collate, reverse_collate)
     local sorting
-    if collate == "strcoll" then
-        sorting = function(a, b)
-            return ffiUtil.strcoll(a.text, b.text)
-        end
-    elseif collate == "natural" then
-        local natsort
-        -- Only keep the cache if we're an *instance* of FileChooser
-        if self ~= FileChooser then
-            natsort, self.natsort_cache = sort.natsort_cmp(self.natsort_cache)
-        else
-            natsort = sort.natsort_cmp()
-        end
-        sorting = function(a, b)
-            return natsort(a.text, b.text)
-        end
-    elseif collate == "access" then
-        sorting = function(a, b)
-            return a.attr.access > b.attr.access
-        end
-    elseif collate == "date" then
-        sorting = function(a, b)
-            return a.attr.modification > b.attr.modification
-        end
-    elseif collate == "size" then
-        sorting = function(a, b)
-            return a.attr.size < b.attr.size
-        end
-    elseif collate == "type" then
-        sorting = function(a, b)
-            if (a.suffix or b.suffix) and a.suffix ~= b.suffix then
-                return ffiUtil.strcoll(a.suffix, b.suffix)
-            end
-            return ffiUtil.strcoll(a.text, b.text)
-        end
-    else -- collate == "percent_unopened_first" or collate == "percent_unopened_last"
-        sorting = function(a, b)
-            if a.opened == b.opened then
-                if a.opened then
-                    return a.percent_finished < b.percent_finished
-                end
-                return ffiUtil.strcoll(a.text, b.text)
-            end
-            if collate == "percent_unopened_first" then
-                return b.opened
-            end
-            return a.opened
-        end
+    if self ~= FileChooser then
```

Comment:
Not entirely sure this is doing what you think it does ;).

(i.e., I'm *fairly* confident FileChooser is not, in fact, a singleton).

Reply:
I overlooked that one; that does jump out as a bit suspicious. ;-P

Reply:
Oh, I see from the previous code that this is on purpose (and I may actually have written that myself ;p), but this just goes to show that murdering comments is bad ;o).

Reply:
+1 on the murdering comments thing

Reply:
Yeah, I copied what you wrote because I couldn't really figure out when FileChooser has an instance and when it does not. I shouldn't have murdered the comment though, but luckily I can revive it. ;)